src/Common/src/System/PasteArguments.cs

```diff
@@ -0,0 +1,144 @@
+                        stringBuilder.Append(Quote);
+                        int idx = 0;
+                        while (idx < argument.Length)
+                        {
+                            char c = argument[idx++];
+                            if (c == Backslash)
+                            {
+                                int numBackSlash = 1;
+                                while (idx < argument.Length && argument[idx] == Backslash)
+                                {
+                                    idx++;
+                                    numBackSlash++;
+                                }
+                                if (idx == argument.Length)
+                                {
+                                    // We'll emit an end quote after this so must double the number of backslashes.
+                                    stringBuilder.Append(Backslash, numBackSlash * 2);
+                                }
+                                else if (argument[idx] == Quote)
+                                {
+                                    // Backslashes will be followed by a quote. Must double the number of backslashes.
+                                    stringBuilder.Append(Backslash, numBackSlash * 2 + 1);
+                                    stringBuilder.Append(Quote);
+                                    idx++;
+                                }
+                                else
+                                {
+                                    // Backslash will not be followed by a quote, so emit as normal characters.
+                                    stringBuilder.Append(Backslash, numBackSlash);
+                                }
+                                continue;
+                            }
+                            if (c == Quote)
+                            {
+                                // Escape the quote so it appears as a literal. This also guarantees that we won't end up generating a closing quote followed
+                                // by another quote (which parses differently pre-2008 vs. post-2008.)
+                                stringBuilder.Append(Backslash);
+                                stringBuilder.Append(Quote);
+                                continue;
+                            }
+                            stringBuilder.Append(c);
+                        }
+                        stringBuilder.Append(Quote);
+                    }
+                }
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        private static bool ContainsNoWhitespaceOrQuotes(string s)
+        {
+            for (int i = 0; i < s.Length; i++)
+            {
+                char c = s[i];
+                if (char.IsWhiteSpace(c) || c == Quote)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private const char Nul = (char)0;
```

Comment:
The "Nul" was left over from a previous 0-terminated version - this isn't used anymore.